### PR TITLE
Feature/40009

### PR DESCRIPTION
**Pull request by UrvashiChaudhary**
- Create a file authMoiddleWare.ts inside libs/routes.
- Install a dependency 'jsonwebtoken',
- Create a middleware 'authMiddleWare' with following:

1. Accepts two arguments(module and permission type).
2. Retrieve token from the header 'Authorization' and decode it using .verify method. It will give the details of user.
3. If no user decoded. Throws an error of 403. Unauthorized Access.
4. Use the function hasPermission defined earlier in utils for checking the permission.
5. If permission does not exists, throws error.
 

- Sample token for various roles are:

eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJTdWNjZXNzaXZlIFRlY2hub2xvZ2llcyIsImlhdCI6MTU0ODY5MDQzNywiZXhwIjoxNTgwNzk3MjU5LCJhdWQiOiJ3d3cuc3VjY2Vzc2l2ZS5pbiIsInN1YiI6IkxlYXJuIGFuZCBJbXBsZW1lbnQiLCJuYW1lIjoiVHJhaW5lZSIsImVtYWlsIjoidHJhaW5lZUBzdWNjZXNzaXZlLnRlY2giLCJyb2xlIjoidHJhaW5lZSIsImp0aSI6ImEwMjBhYTJlLTk1ZDUtNDc5MC05NDAxLWYyMjFlNDBiNjA1MCJ9.-z2lCFkZJaW5ACZurZH2NiAh9VhefjRCvmlDw1X57Rs

{ "iss": "Successive Technologies", "iat": 1548690437, "exp": 1580226437, "aud": "www.successive.in&quot;, "sub": "Learn and Implement", "name": "Trainee", "email": "trainee@successive.tech", "role": "trainee" } key: qwertyuiopasdfghjklzxcvbnm123456

eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJTdWNjZXNzaXZlIFRlY2hub2xvZ2llcyIsImlhdCI6MTU0ODY5MDQzNywiZXhwIjoxNTgwNzk3MzkxLCJhdWQiOiJ3d3cuc3VjY2Vzc2l2ZS5pbiIsInN1YiI6IkxlYXJuIGFuZCBJbXBsZW1lbnQiLCJuYW1lIjoiSGVhZCBUcmFpbmVyIiwiZW1haWwiOiJoZWFkLnRyYWluZXJAc3VjY2Vzc2l2ZS50ZWNoIiwicm9sZSI6ImhlYWQtdHJhaW5lciIsImp0aSI6IjhhMjNkNjc1LWY1NjctNDNmMy1iMDQxLWE3YzlkM2VhMzI4ZCJ9.ciRD8ypgm90pOfD5AD2NKKGOnE7Ke4M0ubSNZvQIaKw { "iss": "Successive Technologies", "iat": 1548690437, "exp": 1580226437, "aud": "www.successive.in&quot;, "sub": "Learn and Implement", "name": "Head Trainer", "email": "head.trainer@successive.tech", "role": "head-trainer" } key: qwertyuiopasdfghjklzxcvbnm123456

### CODE REVIEW

File: src/libs/routes/authMiddleWare.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import * as jwt from 'jsonwebtoken';
+import {hasPermission}  from '../permissions';
+import {permissions} from '../constants'
+export default ( module, permissionType )=> (req,res,next)=>{
+    try{
+    console.log("The config is: ",module,permissionType);
+    console.log("Header is: ", req.headers['authorization']);
+    const token= req.headers['authorization']
+    const decodeUser= jwt.verify(token,'qwertyuiopasdfghjklzxcvbnm123456')
```
Comment (adilsuccessive) [inline src/libs/routes/authMiddleWare.ts]: fetch key through config from env.

File: src/controllers/trainee/routes.ts [hunk 2]
```diff
@@ -5,10 +5,14 @@ import validationHandler from '../../libs/validationHandler';
...
+
 const traineeRouter: Router = Router();
 
 traineeRouter.route('/')
-    .get(validationHandler(validation.get),traineeController.get)
+    .get(authMiddleWare('getUser', 'read'),validationHandler(validation.get),traineeController.get)
+    //.get(validationHandler(validation.get),traineeController.get)
     .post(validationHandler(validation.create),traineeController.create)
     .put(validationHandler(validation.update),traineeController.update)
     .delete(validationHandler(validation.delete),traineeController.update)
```
Comment (adilsuccessive) [inline src/controllers/trainee/routes.ts]: add authMiddleWare in all the routes.

File: src/libs/routes/authMiddleWare.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    console.log('user', decodeUser);
+    const a=decodeUser.Role;
+   hasPermission(permissions.getUsers,a,permissionType);
```
Comment (adilsuccessive) [inline src/libs/routes/authMiddleWare.ts]: authmiddleware is not working as per the task.

**Review by adilsuccessive — CHANGES_REQUESTED** [review]
also fix lint.

File: src/libs/routes/authMiddleWare.ts (hunk 1 above)
Comment (UrvashiChaudhary) [reply src/libs/routes/authMiddleWare.ts]: done

File: src/controllers/trainee/routes.ts (hunk 2 above)
Comment (UrvashiChaudhary) [reply src/controllers/trainee/routes.ts]: done

File: src/libs/routes/authMiddleWare.ts (hunk 3 above)
Comment (UrvashiChaudhary) [reply src/libs/routes/authMiddleWare.ts]: resolved

Comment (UrvashiChaudhary) [reply src/libs/routes/authMiddleWare.ts]: lint fixed

**Pull request merged**